src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,742 @@
+      return bytesPerImage * (copyExtent.depth - 1) + bytesInLastSlice;
+    }
+  }
+
+  // Offset for a particular texel in the linear texture data
+  getTexelOffsetInBytes(
+    textureDataLayout: GPUTextureDataLayout,
+    format: SizedTextureFormat,
+    texel: Required<GPUOrigin3DDict>,
+    origin: Required<GPUOrigin3DDict> = { x: 0, y: 0, z: 0 }
+  ): number {
+    const { offset, bytesPerRow, rowsPerImage } = textureDataLayout;
+    const info = kSizedTextureFormatInfo[format];
+
+    assert(texel.x >= origin.x && texel.y >= origin.y && texel.z >= origin.z);
+    assert(rowsPerImage! % info.blockHeight! === 0);
+    assert(texel.x % info.blockWidth! === 0);
+    assert(texel.y % info.blockHeight! === 0);
+    assert(origin.x % info.blockWidth! === 0);
+    assert(origin.y % info.blockHeight! === 0);
+
+    const bytesPerImage = (rowsPerImage! / info.blockHeight!) * bytesPerRow;
+
+    return (
+      offset! +
+      (texel.z - origin.z) * bytesPerImage +
+      ((texel.y - origin.y) / info.blockHeight!) * bytesPerRow +
+      ((texel.x - origin.x) / info.blockWidth!) * info.bytesPerBlock!
+    );
+  }
+
+  *iterateBlockRows(
+    size: GPUExtent3DDict,
+    origin: Required<GPUOrigin3DDict>,
+    format: SizedTextureFormat
+  ): Generator<Required<GPUOrigin3DDict>> {
+    const info = kSizedTextureFormatInfo[format];
+    assert(size.height % info.blockHeight! === 0);
+    for (let y = 0; y < size.height / info.blockHeight!; ++y) {
+      for (let z = 0; z < size.depth; ++z) {
+        yield {
+          x: origin.x,
+          y: origin.y + y * info.blockHeight!,
+          z: origin.z + z,
+        };
+      }
+    }
+  }
+
+  generateData(byteSize: number): Uint8Array {
+    const arr = new Uint8Array(byteSize);
+    for (let i = 0; i < byteSize; ++i) {
+      arr[i] = (i ** 3 + i) % 251;
+    }
+    return arr;
+  }
+
+  // Put data into texture with an appropriate method.
+  initTexture(
+    textureCopyView: GPUTextureCopyView,
+    textureDataLayout: GPUTextureDataLayout,
+    copySize: GPUExtent3D,
+    partialData: Uint8Array,
+    method: string
+  ): void {
+    switch (method) {
+      case 'WriteTexture': {
+        this.device.defaultQueue.writeTexture(
+          textureCopyView,
+          partialData,
+          textureDataLayout,
+          copySize
+        );
+
+        break;
+      }
+      case 'CopyB2T': {
+        const buffer = this.device.createBuffer({
+          mappedAtCreation: true,
+          size: align(partialData.byteLength, 4),
+          usage: GPUBufferUsage.COPY_SRC,
+        });
+        new Uint8Array(buffer.getMappedRange()).set(partialData);
+        buffer.unmap();
+
+        const encoder = this.device.createCommandEncoder();
+        encoder.copyBufferToTexture({ buffer, ...textureDataLayout }, textureCopyView, copySize);
+        this.device.defaultQueue.submit([encoder.finish()]);
+
+        break;
+      }
+      default:
+        unreachable();
+    }
+  }
+
+  checkData(
```

Comment:
nit: It could help readability to have comments explaining what these functions do, and highlight the different between checkData and fullCheck, ditto for other largish functions.